src/server/scripts/Pandaria/TerraceOfEndlessSpring/boss_protectors_of_the_endless.cpp

```diff
@@ -0,0 +1,766 @@
+            Talk(SAY_CORRUPTED_ORB);
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
+
+        events.Update(diff);
+
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
+
+        if (Aura* shaCorruption = me->GetAura(SPELL_SHA_CORRUPTION))
+        {
+            if (shaCorruption->GetStackAmount() >= 1 && !_newPower)
+            {
+                events.ScheduleEvent(EVENT_CORRUPTED_WATERS, 5100ms);
+                _newPower = true;
+            }
+            else if (shaCorruption->GetStackAmount() == 2 && !_fullPower)
+            {
+                Talk(SAY_FULL_POWER);
+                DoCastSelf(SPELL_OVERWHELMING_CORRUPTION);
+                _fullPower = true;
+            }
+        }
+
+        while (uint32 eventId = events.ExecuteEvent())
+        {
+            ExecuteEvent(eventId);
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+        }
+    }
+};
+
+// 122851 - Raid Warning: I'm Standing In Cleansing Waters
+class spell_protectors_of_the_endless_warning_cleansing_waters : public SpellScript
+{
+    void HandleHit(SpellEffIndex /*effIndex*/)
+    {
+        if (Creature* protector = GetCaster()->ToCreature())
+            if (protector->GetEntry() == BOSS_PROTECTOR_KAOLAN || protector->GetEntry() == BOSS_ELDER_ASANI || protector->GetEntry() == BOSS_ELDER_REGAIL)
+                protector->AI()->Talk(SAY_WARNING_CLEANSING_WATERS, GetHitUnit()->ToPlayer());
+    }
+
+    void Register() override
+    {
+       OnEffectHitTarget += SpellEffectFn(spell_protectors_of_the_endless_warning_cleansing_waters::HandleHit, EFFECT_0, SPELL_EFFECT_DUMMY);
+    }
+};
+
+// 111850 - Lightning Prison Marker
+class spell_protectors_of_the_endless_lightning_prison_marker : public SpellScript
+{
+    void FilterTargets(std::list<WorldObject*>& targets)
+    {
+        Trinity::Containers::RandomResize(targets, GetCaster()->GetMap()->Is25ManRaid() ? 3 : 2);
+    }
+
+    void Register()
+    {
+        OnObjectAreaTargetSelect += SpellObjectAreaTargetSelectFn(spell_protectors_of_the_endless_lightning_prison_marker::FilterTargets, EFFECT_0, TARGET_UNIT_SRC_AREA_ENEMY);
+    }
+};
+
+// 122874 - Lightning Prison Cast
+class spell_protectors_of_the_endless_lightning_prison_cast : public SpellScript
+{
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo({ SPELL_LIGHTNING_PRISON_MARKER });
+    }
+
+    void HandleCast()
+    {
+        GetCaster()->CastSpell(GetCaster(), SPELL_LIGHTNING_PRISON_MARKER, true);
+    }
+
+    void Register() override
+    {
+        AfterCast += SpellCastFn(spell_protectors_of_the_endless_lightning_prison_cast::HandleCast);
+    }
+};
+
+// 117052 - Sha Corruption
+class spell_protectors_of_the_endless_sha_corruption : public SpellScript
+{
+    void HandleScript(SpellEffIndex /*effIndex*/)
+    {
+        GetHitUnit()->SetFullHealth();
```

Comment:
you might want to filter targets first to only include the bosses and not the caster itself, otherwise you'd be healing literally everyone here

Reply:
i filtered with conditions

Reply:
has SPELL_ATTR1_EXCLUDE_CASTER, this is fine